### ISSUE TITLE
Enforce ReadWriteOnce PVC access mode during scheduling

### ISSUE THREAD

**Issue by chrishenzie** [kubernetes/kubernetes#103305]
**What would you like to be added:**
The ReadWriteOnce PVC access mode guarantees only the pods on a single node can access the provided PVC as a volume. This is enforced during volume attachment (in kube-controller-manager), and should be additionally enforced during scheduling.

**Why is this needed:**
Because without it the scheduler will assign pods to nodes and they will fail to start (due to failed volume attachment).

**Expected behavior:**
When pod 1 is scheduled, if it uses a `ReadWriteOnce` PVC that's in use by a pod 2 running on node 1, then pod 1 can only be scheduled to node 1.

/sig scheduling
/sig storage
/cc @alculquicondor 
/cc @jsafrane

**Comment by alculquicondor** [comment]
/triage accepted
@chrishenzie can you work on it before the code freeze?

**Comment by chrishenzie** [comment]
I don't have bandwidth to work on this before code freeze. I will need to sync with sig storage to find an owner for this, likely in a following release.

**Comment by alculquicondor** [comment]
No worries

/help

**Comment by chrishenzie** [comment]
Before this is implemented, we need to spend some time evaluating any potential issues with this feature. If there are any backwards compatibility issues, or customers depending on the legacy (broken) behavior. I will present this to SIG storage in our next meeting to gather feedback.

**Comment by bswartz** [comment]
I'm concerned that any changes here might not be backwards compatible. As crazy as the current behavior is (create a RWO PVC, create 2 pods that use it, let them schedule to different nodes, and the second pod waits for the first to die) it's far from impossible that someone depends on that behavior, and to change it could conceivably break them.
It's already possible to ensure 2 pods that intend to share access to a single RWO PVC end up on the same node using affinities. I don't feel like anything additional is needed.

**Comment by alculquicondor** [comment]
Once the second Pod is scheduled to a different node, doesn't kubelet mark it as failed?

Just letting these pods continuously scheduled/failed/recreated is something I would prefer we avoid.

**Comment by bswartz** [comment]
The second pod just gets stuck in pending while the attachment fails in a loop. When the first pod eventually dies, the attachments stop failing and it's able to start. Not sure who would desire this, but the behavior is reliable.

**Comment by alculquicondor** [comment]
Uhm... in that case, the change wouldn't be backwards compatible. Thanks for flagging, Ben.

**Comment by chrishenzie** [comment]
Quick thought experiment on how this would work depending on the affinity / anti-affinity rules used.

If a user has two pods with anti-affinity rules and sharing the same ReadWriteOnce PVC, one will start and the other will get stuck waiting for volume attachment. If we were to enforce ReadWriteOnce in the scheduler in this case, the second pod would fail to schedule because there's no place to go (there would only be one viable node but it's occupied by the other pod).

In this scenario I think it would be okay to introduce this behavior change. In either situation the pod would fail to start, it's just a matter of where it fails (the AttachDetachController vs. the scheduler). Unless customers are somehow consuming the specific failure modes on these pods I think it would be okay to do.

However, this will cause big issues for some workloads if they aren't using anti-affinity rules.

Let's say a customer has a workload that requires it be the only writer for their ReadWriteOnce PVC. They run more than one pod (even though they probably shouldn't), but it's been alright so far because they have a cluster with 1000s of nodes and so pods usually always end up on different nodes, so only one ever starts.

In this scenario it's definitely not okay to introduce the behavior change. If we start enforcing ReadWriteOnce in the scheduler and more than one pod lands on the same node and starts writing this will cause issues for the customer.

**Comment by kerthcet** [comment]
sorry for interruption, `ReadWriteOnce` is ambiguous. 🥲

**Comment by alculquicondor** [comment]
@liggitt for an API compatibility perspective

**Comment by kerthcet** [comment]
I do agree with @chrishenzie, especially to new comers, they have less knowledge about affinity, they would treat this behavior as a bug or misconfiguration. My poor opinion.🧐

**Comment by chrishenzie** [comment]
/cc @thockin

After further review it doesn't look like there is a backwards compatible way of enforcing ReadWriteOnce in the scheduler.

In API review it was suggested we aim for a consistent approach (using scheduler for both access modes or neither). Since it seems we won't be able to use the scheduler to enforce ReadWriteOnce, are we able to make an exception and still allow for the scheduler to enforce ReadWriteOncePod? Doing so allows us to fail faster

**Comment by alculquicondor** [comment]
I think inconsistency in this case is acceptable, given that not enforcing ReadWriteOnce in scheduler was probably an oversight.

**Comment by bswartz** [comment]
Whether it was an oversight depends on what you thought it was supposed to do at the time. There was fundamental confusion about that point which led to the need for this new access mode. Those who thought that RWO meant "one pod" would have assumed there was no work for the scheduler to do. Those who thought that it meant "one node" were probably content with the options available to them at the time, although a scheduler change might have been viewed by them as a bonus. A similar confusion occurred in the CSI spec, and for reasons that I won't go into, the CSI spec ended up being exactly backwards of what Kuberentes did, which necessitated this change to sort out the mess.

**Comment by boenn** [comment]
> /cc @thockin
> 
> After further review it doesn't look like there is a backwards compatible way of enforcing ReadWriteOnce in the scheduler.
> 
> In API review it was suggested we aim for a consistent approach (using scheduler for both access modes or neither). Since it seems we won't be able to use the scheduler to enforce ReadWriteOnce, are we able to make an exception and still allow for the scheduler to enforce ReadWriteOncePod? Doing so allows us to fail faster

+1 ，I think the difference between rwop and rwo is that if there are multiple pods on the same node in the former, only one can successfully mount the pvc, and the status of the other pods is failed or pending (preemption may occur); while the latter can be successfu mount and run. Pods on the same node use pvc in a certain order (according to priority or other) may be a feature of rwop, but for rwo it is on different nodes, which is different.



By the way, the name rwo is really confusing, it's impossible to tell whether the latitude is ''node" or "pod" from the name.

**Comment by boenn** [comment]
> /cc @thockin
> 
> After further review it doesn't look like there is a backwards compatible way of enforcing ReadWriteOnce in the scheduler.
> 
> In API review it was suggested we aim for a consistent approach (using scheduler for both access modes or neither). Since it seems we won't be able to use the scheduler to enforce ReadWriteOnce, are we able to make an exception and still allow for the scheduler to enforce ReadWriteOncePod? Doing so allows us to fail faster

+1 ，I think the difference between rwop and rwo is that if there are multiple pods on the same node in the former, only one can successfully mount the pvc, and the status of the other pods is failed or pending (preemption may occur); while the latter can be successfu mount and run. Pods on the same node use pvc in a certain order (according to priority or other) may be a feature of rwop, but for rwo it is on different nodes, which is different.



By the way, the name rwo is really confusing, it's impossible to tell whether the latitude is ''node" or "pod" from the name.

**Comment by bswartz** [comment]
Yes it is confusing, and confusion in the past led us to make decisions that we now regret, and the addition of the RWOP access mode is an attempt to clean up the mess and reduce the confusion. RWO has always meant "one node" in practice, despite assumption to the contrary. Docs could be cleaned up, but the API/UI will remain backwards compatible.

**Comment by liggitt** [comment]
> Let's say a customer has a workload that requires it be the only writer for their ReadWriteOnce PVC. They run more than one pod without anti-affinity rules (even though they probably shouldn't), but it's been alright so far because they have a cluster with 1000s of nodes and so pods usually always end up on different nodes, so only one ever starts.

"usually always end up on different nodes" doesn't sound like "alright so far"... if their pods shouldn't access the volume simultaneously, and they're not actively *preventing* the pods from landing on the same nodes using anti-affinity, they are non-deterministically broken, right?

> In this scenario it's definitely not okay to introduce the behavior change. If we start enforcing ReadWriteOnce in the scheduler and more than one pod lands on the same node and starts writing this will cause issues for the customer when both of them start writing simultaneously.

Why not? Aren't they already exposed to that bug in a non-deterministic way?

What do you mean by "enforcing ReadWriteOnce in the scheduler"?

**Comment by chrishenzie** [comment]
> if their pods shouldn't access the volume simultaneously, and they're not actively preventing the pods from landing on the same nodes using anti-affinity, they are non-deterministically broken, right?

Yes

> Aren't they already exposed to that bug in a non-deterministic way?

Yes they would be. My concern was around how customers may depend on this buggy behavior, and how fixing it could cause issues for some workloads that are non-deterministically broken.

> What do you mean by "enforcing ReadWriteOnce in the scheduler"?

If more than one pod references a PVC with ReadWriteOnce, the scheduler will now force them to land on the same node to ensure they all have access (whereas previously they will land wherever, and the first to attach the PVC wins while the others sit around trying to attach the volume if they land on different nodes).

**Comment by liggitt** [comment]
> > Aren't they already exposed to that bug in a non-deterministic way?
> 
> Yes they would be. My concern was around how customers may depend on this buggy behavior, and how fixing it could cause issues for some workloads that are non-deterministically broken.

IMO, taking something that was non-deterministically broken and making it deterministically broken is in-bounds (and actually a net improvement, since it makes issues show up more consistently during test/dev), but I'm happy to hear other perspectives on it.

> If more than one pod references a PVC with ReadWriteOnce, the scheduler will now force them to land on the same node to ensure they all have access (whereas previously they will land wherever, and the first to attach the PVC wins while the others sit around trying to attach the volume if they land on different nodes).

If this was done as yet another filter/prioritization step by the scheduler, that seems reasonable at first glance. If the user specified affinity/anti-affinity constraints that conflicted with the PVC-access-mode-derived constraints, you could end in situations where some pods were unschedulable, but from the perspective of container processes or pod readiness, a pod scheduled to a node that cannot start, and a pod that cannot be scheduled to a node because of constraints are pretty similar...

It is easy to imagine combinations of pods / pvcs / disruption budgets / node states that would cause problems ("cannot drain node N because we cannot kill pod P because we cannot disrupt the pods for deployment D but we cannot schedule replacement pod P2 to a different node because the scheduler will only schedule it to node N which we're trying to drain and have made unschedulable", etc). If the scheduler *preferred* placing pods with ReadWriteOnce PVC volumes on nodes already running pods using those PVCs but did not hard-fail, that might be a better starting point.

**Comment by alculquicondor** [comment]
> It is easy to imagine combinations of pods / pvcs / disruption budgets / node states that would cause problems ("cannot drain node N because we cannot kill pod P because we cannot disrupt the pods for deployment D but we cannot schedule replacement pod P2 to a different node because the scheduler will only schedule it to node N which we're trying to drain and have made unschedulable", etc).

Correct me if I'm wrong, but I thought draining and eviction doesn't check if the Pod would be schedulable somewhere else, only if there are other replicas running, as required by the PDB).

On the other hand, Cluster Autoscaler does indeed simulate. But I would image that it would take out the candidate node for scale-down out of the simulation. @MaciekPytel does this sound reasonable?

**Comment by kerthcet** [comment]
> > It is easy to imagine combinations of pods / pvcs / disruption budgets / node states that would cause problems ("cannot drain node N because we cannot kill pod P because we cannot disrupt the pods for deployment D but we cannot schedule replacement pod P2 to a different node because the scheduler will only schedule it to node N which we're trying to drain and have made unschedulable", etc).
> 
> Correct me if I'm wrong, but I thought draining and eviction doesn't check if the Pod would be schedulable somewhere else, only if there are other replicas running, as required by the PDB).
> 
> On the other hand, Cluster Autoscaler does indeed simulate. But I would image that it would take out the candidate node for scale-down out of the simulation. @MaciekPytel does this sound reasonable?

confused + 1, like hard affinity, if all the pods scheduled in a single node and you drain the node ,all the new created pods will pending until all the old ones terminated successfully.  The service will hang up for a while although. (but PDB may block the process and make it unschedulable)

**Comment by kerthcet** [comment]
> > It is easy to imagine combinations of pods / pvcs / disruption budgets / node states that would cause problems ("cannot drain node N because we cannot kill pod P because we cannot disrupt the pods for deployment D but we cannot schedule replacement pod P2 to a different node because the scheduler will only schedule it to node N which we're trying to drain and have made unschedulable", etc).
> 
> Correct me if I'm wrong, but I thought draining and eviction doesn't check if the Pod would be schedulable somewhere else, only if there are other replicas running, as required by the PDB).
> 
> On the other hand, Cluster Autoscaler does indeed simulate. But I would image that it would take out the candidate node for scale-down out of the simulation. @MaciekPytel does this sound reasonable?

confused + 1, like hard affinity, if all the pods scheduled in a single node(not recommended but can be achieved) and you drain the node ,all the new created pods will pending until all the old ones terminated successfully.  The service will hang up for a while although. (but PDB may block the process and make it unschedulable)

**Comment by kerthcet** [comment]
> > It is easy to imagine combinations of pods / pvcs / disruption budgets / node states that would cause problems ("cannot drain node N because we cannot kill pod P because we cannot disrupt the pods for deployment D but we cannot schedule replacement pod P2 to a different node because the scheduler will only schedule it to node N which we're trying to drain and have made unschedulable", etc).
> 
> Correct me if I'm wrong, but I thought draining and eviction doesn't check if the Pod would be schedulable somewhere else, only if there are other replicas running, as required by the PDB).
>
 confused + 1, like hard affinity, if all the pods scheduled in a single node and you drain the node ,all the new created pods will pending until all the old ones terminated successfully. The service will hang up for a while although. (but PDB may block the process and make it unschedulable)

**Comment by liggitt** [comment]
I think you're right, disruption budget / eviction was probably the wrong example. Are there scenarios where an unhealthy node would result in its volumes getting detached, but still leave terminating pods present in the API until the kubelet on the unhealthy node acks/deletes them?

**Comment by chrishenzie** [comment]
/remove-lifecycle stale

I am working to move this feature to beta

**Comment by kerthcet** [comment]
Revisit the problem as we met this in our customer's clusters recently, brother pods with `ReadWriteOnce` PVC in different nodes will stuck in `ContainerCreating`, this is not what we expected. And actually I can't think of a reasonable scenario depending on this behavior. On the opposite, it wastes the resources.
/assign

**Comment by chrishenzie** [comment]
Thank you for taking a look @kerthcet. For context, this is a [beta graduation criteria item](https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/2485-read-write-once-pod-pv-access-mode#beta) for the `ReadWriteOncePod` feature. I had plans on tackling this but if you would like to take a look please do.

**Comment by jsafrane** [comment]
@kerthcet / @chrishenzie Hello, did you have any chance to look at this?

**Comment by kerthcet** [comment]
It seems like @chrishenzie is following this. https://github.com/kubernetes/enhancements/pull/3537

**Comment by jsafrane** [comment]
I can't find any proposed change in ReadWriteOnce access mode scheduling in 
 https://github.com/kubernetes/enhancements/pull/3537, despite the [discussion above](https://github.com/kubernetes/kubernetes/issues/103305#issuecomment-1180949155) says otherwise. @chrishenzie can you please point me to the right section in the KEP?

/unassign @kerthcet 
/unassign @yuzhiquan
/assign @chrishenzie

**Comment by chrishenzie** [comment]
@jsafrane It's mentioned in the [beta graduation criteria](https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/2485-read-write-once-pod-pv-access-mode/README.md#beta) but not explicitly called out. I can add a section for it.

For https://github.com/kubernetes/enhancements/pull/3537, we're going to have to delay beta graduation until 1.27 due to missing E2E tests for alpha. I'm currently prototyping some and plan to have a PR soon.

**Comment by lixd** [comment]
hi,is there any progress about this issue？i see there is a new access mode [RWOP](https://github.com/kubernetes/enhancements/pull/3537),so for non-attachable volume is better to use ROWP to replace RWO? if that how to deal with this scene: share volume in a single node?

**Comment by chrishenzie** [comment]
@jsafrane Apologies, I'm getting this issue confused with #103132 (`ReadWriteOncePod` vs. `ReadWriteOnce`). This issue is not tracked yet as part of the KEP and no progress to report.

@lixd You can use `ReadWriteOncePod` if you only want a single pod on the node to access the volume. If you want to share the volume with multiple pods on the same node then you can use `ReadWriteOnce`, but you will probably need to use [pod affinity](https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#inter-pod-affinity-and-anti-affinity) to make sure they are located together.

**Comment by lixd** [comment]
@chrishenzie thanks for your replay

**Comment by AxeZhan** [comment]
What's the state of this issue? Do we still need to enforce this in scheduling? :)

**Comment by travisghansen** [comment]
Yes, currently volumeattchments are required to enforce proper rwo semantics.

**Comment by AxeZhan** [comment]
> Yes, currently volumeattchments are required to enforce proper rwo semantics.

However, I don't see a conclusion in before discussions :). We were talking about RWOP instead of RWO after some certain time.  @kerthcet Do you think we still need this in scheduler? I can help if we still need this.

**Comment by kerthcet** [comment]
I guess nothing left right now, it's already beta. cc @chrishenzie for confirmation, the last one about scheduling is https://github.com/kubernetes/kubernetes/pull/114051 IIRC.

**Comment by AxeZhan** [comment]
> I guess nothing left right now, it's already beta. cc @chrishenzie for confirmation, the last one about scheduling is #114051 IIRC.

The prs involved so far is all about RWOP, not RWO. See https://github.com/kubernetes/kubernetes/issues/103305#issuecomment-1266575866 https://github.com/kubernetes/kubernetes/issues/103305#issuecomment-1287210083 also

**Comment by windsource** [comment]
It would be great if RWO could be handled by the scheduler as well. As a workaround I thought about creating a mutating web hook in order to set the pod affinty and not have to do it manually but somehow this feels odd. Would be much cleaner if the scheduler handles this and schedules pods on the same node.

**Comment by chrishenzie** [comment]
Hi -- nothing is left for `ReadWriteOncePod` except for the bump to GA next release. The scheduler enforcing `ReadWriteOnce` is a separate topic, apologies for the confusion. I don't have the cycles to take this on right now, but anyone else is free to take it. There will probably need to be some followup discussions in the Kubernetes CSI / SIG storage meetings to track this and determine how to handle this [non-backwards compatible behavior change](https://github.com/kubernetes/kubernetes/issues/103305#issuecomment-885095233).

> Is RWOP created to solve this issue?

No, RWOP was created to handle the case where only a single Pod should have access. `ReadWriteOnce` is different in that multiple Pods on the same Node can access the storage.
